server/common/config/base_config.py

```diff
@@ -0,0 +1,104 @@
+import copy
+
+from flatten_dict import flatten
+from server.common.errors import ConfigurationError
+
+
+class BaseConfig(object):
+    """This class handles the mechanics of updating and checking attributes.
+    Derived classes are expected to store the actual attributes"""
+
+    def __init__(self, app_config, default_config, dictval_cases={}):
+        # reference back to the app_config
+        self.app_config = app_config
+        # the complete set of attribute and their default values (unflattened)
+        self.default_config = default_config
+        # attributes where the value may be a dict (and therefore are not flattened)
+        self.dictval_cases = dictval_cases
```

Comment:
Why does this need to be a separate member variable? During flattening time, could you not just check one variable and based on the type, triage whether it needs to be flattened or not. The the flattening logic isn't sprawled beyond the flattening function itself.

Reply:
The whole config is passed in as a dict of dicts. Most of the time the config value is a a string, but when the value is intended to be a dict we dont want to flatten it and turn the dict keys into the path. @bmccandless does this align with your understanding?

Reply:
Yes, that's right.   These attributes are generally for interfacing to external modules.  One example is tiledb context parameters.   We don't processes them in cellxgene, but rather pass them (as a dictionary of key/value) to a tiledb interface.